### PR TITLE
i150: Migrate to Navigator2.0

### LINKED ISSUE #150: Migrate to Navigator2.0 - step 1 (pages)
- [ ] create a `NavigationPage` model to represent a [Page](https://api.flutter.dev/flutter/widgets/Navigator/pages.html)
- [ ] add `navigationPages` to app state 
- [ ] add an extension to convert `NavigationPage` to `Page` 
- [ ] add a store connector above the material app widget that rebuilds on changes to `navigationPages` 
  - [ ] in converter, convert NavigationPages to Pages and update `pages` member of Navigation  member of MaterialApp widget 
- [ ] add an action for each instance where we used the Navigator to push/pop 
  - [ ] replace all Navigator1 imperative calls with actions 
- [ ] add a reducer to respond to each action

### PR DESCRIPTION
**Pull request by nickmeinhold**
fixes #150

### CODE REVIEW

File: lib/actions/navigation/push_page.dart [hunk 1]
```diff
@@ -0,0 +1,34 @@
+library push_page;
```
Comment (nickmeinhold) [inline lib/actions/navigation/push_page.dart:1]: there are just two actions now for all our navigation: PushPage and RemoveCurrentPage

File: lib/actions/navigation/remove_current_page.dart [hunk 2]
```diff
@@ -0,0 +1,30 @@
+library remove_current_page;
```
Comment (nickmeinhold) [inline lib/actions/navigation/remove_current_page.dart:1]: there are just two actions now for all our navigation: PushPage and RemoveCurrentPage

File: lib/extensions/extensions.dart [hunk 3]
```diff
@@ -1,10 +1,11 @@
-export 'firebase_user_extensions.dart';
```
Comment (nickmeinhold) [inline lib/extensions/extensions.dart:1]: this is just the file that allows easy import of all extensions

File: lib/extensions/models_extensions.dart [hunk 4]
```diff
@@ -0,0 +1,41 @@
+import 'package:built_collection/built_collection.dart';
```
Comment (nickmeinhold) [inline lib/extensions/models_extensions.dart:1]: the purpose of this extension is to keep our models as PODOs by keeping the conversion of PageData objects (PODOs we keep in our app state) to Pages (part of the Flutter framework) 

There is a PageData subclass for each page we navigate to - we keep a list of PageData objects in our app state and have a StoreConnector above the MaterialApp so that when the list changes, the MaterialApp's navigator is updated (ie. creates with a new list of pages) 

So basically we need to convert from PageData to Page, and PageData shouldn't have a dependency on Flutter so we do the conversion with an extension. 

The reason it's done in a weird way (ie. by keeping a separate map) is I found it challenging with the combination of extension functions and built_value classes and this was the best I could come up with. There may well be a much better way.

File: lib/middleware/conversations/create_conversation.dart [hunk 5]
```diff
@@ -1,6 +1,6 @@
 import 'package:crowdleague/actions/conversations/create_conversation.dart';
```
Comment (nickmeinhold) [inline lib/middleware/conversations/create_conversation.dart:1]: Just changing from old navigation strategy to using new navigator

File: lib/models/app/app_state.dart [hunk 6]
```diff
@@ -16,6 +16,8 @@ import 'package:crowdleague/models/conversations/conversation/vm_conversation_pa
 import 'package:crowdleague/models/conversations/new_conversation/vm_new_conversation_page.dart';
```
Comment (nickmeinhold) [inline lib/models/app/app_state.dart:16]: Not much change to app state, just the list of page data

File: lib/models/navigation/page_data/email_auth_page_data.dart [hunk 7]
```diff
@@ -0,0 +1,26 @@
+library email_auth_page_data;
```
Comment (nickmeinhold) [inline lib/models/navigation/page_data/email_auth_page_data.dart:1]: Each page we want to navigate to has a corresponding page data object that can be added to the app state's pagesData member, which in turn is used to create the list of pages that is passed in to the navigator

File: lib/models/navigation/page_data/initial_page_data.dart [hunk 8]
```diff
@@ -0,0 +1,26 @@
+library initial_page_data;
```
Comment (nickmeinhold) [inline lib/models/navigation/page_data/initial_page_data.dart:1]: the widget here (ie. the initial page) is actually called AuthOrMain but InitialPage seemed cleaner - maybe I should have renamed the widget... I'll see if it's a big change, there's already quite a few files

File: lib/models/navigation/page_data/page_data.dart [hunk 9]
```diff
@@ -0,0 +1,13 @@
+library page_data;
```
Comment (nickmeinhold) [inline lib/models/navigation/page_data/page_data.dart:1]: This is the parent class of all the page data's.  Polymorphism in built_value is a bit weird, there may be a better way

File: lib/reducers/app_reducer.dart [hunk 10]
```diff
@@ -14,9 +14,11 @@ import 'package:crowdleague/reducers/conversations/update_conversations_page.dar
 import 'package:crowdleague/reducers/conversations/update_new_conversations_page.dart';
```
Comment (nickmeinhold) [inline lib/reducers/app_reducer.dart:14]: just added a couple of reducers

File: lib/reducers/auth/store_user.dart [hunk 11]
```diff
@@ -1,11 +1,18 @@
+import 'package:built_collection/built_collection.dart';
```
Comment (nickmeinhold) [inline lib/reducers/auth/store_user.dart:1]: this was a fix for a bug where the page on top of the InitialPage does not pop after sign in

File: lib/widgets/crowd_league_app.dart [hunk 12]
```diff
@@ -97,25 +95,35 @@ class _CrowdLeagueAppState extends State<CrowdLeagueApp> {
...
-              routes: <String, WidgetBuilder>{
-                '/other_auth_options': (context) => OtherAuthOptionsPage(),
-                '/email_auth_options': (context) => EmailAuthOptionsPage(),
-                '/conversation': (context) => MessagesPage(),
-                '/new_conversation': (context) => NewConversationPage(),
-                '/profile': (context) => ProfilePage()
-              },
-            );
-          }),
+        distinct: true,
```
Comment (nickmeinhold) [inline lib/widgets/crowd_league_app.dart:98]: There's actually not much changed here - the changes start from `home:` 
- there's an extra StoreConnector that rebuilds on changes to `state.pagesData` 
- the MaterialApp now takes a Navigator, which
  - takes a list of pages, which we build from `state.pagesData` 
  - also takes a callback that fires when the system asks to pop (eg. when the back button is tapped) which we use to update our app state 
- I removed the routes, so I could easily find where I needed to update the navigation - there's a lot more cleaning up to do but I wanted to make it a separate PR as this was already large

File: .github/workflows/ci_cd.yml [hunk 13]
```diff
@@ -37,7 +37,7 @@ jobs:
     - name: Install Flutter tool
       uses: subosito/flutter-action@v1
       with:
-        channel: 'dev' # or: 'stable' or 'beta'
+        channel: 'stable' # or: 'stable' or 'beta'
```
Comment (nickmeinhold) [inline .github/workflows/ci_cd.yml:40]: `dev` doesn't have the latest navigator changes yet, `master` didn't work (the flutter action errors when trying to download) but I read that they had cherry picked the changes onto `stable` so I tried and all tests pass.

**Comment by nickmeinhold** [comment]
Created as separate issues to keep the size of the PR down: 
- Clean out old navigation stuff #155 
- Add tests for navigation #160

File: .github/workflows/ci_cd.yml (hunk 13 above)
Comment (tamari-gray) [reply .github/workflows/ci_cd.yml:40]: ooh stable :0

Comment (tamari-gray) [reply .github/workflows/ci_cd.yml:40]: thanks for the heads up

File: lib/actions/navigation/push_page.dart (hunk 1 above)
Comment (tamari-gray) [reply lib/actions/navigation/push_page.dart:1]: sweet

File: lib/actions/navigation/remove_current_page.dart (hunk 2 above)
Comment (tamari-gray) [reply lib/actions/navigation/remove_current_page.dart:1]: sweeeet :)

File: lib/extensions/extensions.dart (hunk 3 above)
Comment (tamari-gray) [reply lib/extensions/extensions.dart:1]: thanks 🌮

File: lib/middleware/conversations/create_conversation.dart (hunk 5 above)
Comment (tamari-gray) [reply lib/middleware/conversations/create_conversation.dart:1]: navigator

File: lib/models/app/app_state.dart (hunk 6 above)
Comment (tamari-gray) [reply lib/models/app/app_state.dart:16]: thanks :)

File: lib/models/navigation/page_data/initial_page_data.dart (hunk 8 above)
Comment (tamari-gray) [reply lib/models/navigation/page_data/initial_page_data.dart:1]: lol thanks for the heads up, Initial page is really clear to me

File: lib/models/navigation/page_data/email_auth_page_data.dart (hunk 7 above)
Comment (tamari-gray) [reply lib/models/navigation/page_data/email_auth_page_data.dart:1]: ahhh thank you

File: lib/models/navigation/page_data/page_data.dart (hunk 9 above)
Comment (tamari-gray) [reply lib/models/navigation/page_data/page_data.dart:1]: cool got it

File: lib/reducers/app_reducer.dart (hunk 10 above)
Comment (tamari-gray) [reply lib/reducers/app_reducer.dart:14]: sweet

File: lib/reducers/auth/store_user.dart (hunk 11 above)
Comment (tamari-gray) [reply lib/reducers/auth/store_user.dart:1]: and pop she will, nice!

File: lib/utils/serializers.dart [hunk 14]
```diff
@@ -125,6 +128,7 @@ part 'serializers.g.dart';
   PickProfilePic,
   PlatformType,
   PlumbDatabaseStream,
+  PushPage,
   PrintFCMToken,
   RecordAddedRouteInfo,
   RecordRemovedRouteInfo,
```
Comment (tamari-gray) [inline lib/utils/serializers.dart:134]: does removeCurrentPage need to be here?

File: lib/widgets/crowd_league_app.dart (hunk 12 above)
Comment (tamari-gray) [reply lib/widgets/crowd_league_app.dart:98]: sweet thanks for the detailed comment @nickmeinhold!

File: lib/extensions/models_extensions.dart (hunk 4 above)
Comment (tamari-gray) [reply lib/extensions/models_extensions.dart:1]: awesome, after i went through the pr and came back to this it started making sense. the detailed comments helped so much. thanks @nickmeinhold!!!

**Review by tamari-gray — DISMISSED** [review]
great pr, it felt like a shortcut for me to learn how the navigator works :)
And the comments really saved my life, they were my map in the uncharted navigator 2.0 territory. thanks heaps!

there was one comment where i think an action was missed in the serialiser, but im not sure if thats a valid thing to pick up so please merge if its nothing!

File: lib/middleware/conversations/create_conversation.dart (hunk 5 above)
Comment (nickmeinhold) [reply lib/middleware/conversations/create_conversation.dart:1]: hahahaha I loved this movie!!!

File: lib/utils/serializers.dart (hunk 14 above)
Comment (nickmeinhold) [reply lib/utils/serializers.dart:135]: Yes indeed!  Great pickup - thanks!  I've added it now.
